### ISSUE TITLE
[steamshovel] test failure - race condition (Trac #1719)

### ISSUE THREAD

**Issue by icecube-trac** [icecube-trac/tix2#1711]
There is apparently a race condition that needs fixing:

```text
dschultz@tide2:~/Documents/offline_software/serialization/build_1.60$ python steamshovel/resources/resources/test/test_shovelio_against_I3Reader.py 
INFO (I3Module): Opened file /cvmfs/icecube.opensciencegrid.org/data/i3-test-data/sim/GCD.i3.gz (I3Reader.cxx:180 in void I3Reader::OpenNextFile())
Configure with filenamelist: /cvmfs/icecube.opensciencegrid.org/data/i3-test-data/sim/GCD.i3.gz /cvmfs/icecube.opensciencegrid.org/data/i3-test-data/sim/corsika.F2K010001_IC59_slim.i3.gz
frame 1 Geometry
frame 2 Calibration
frame 3 DetectorStatus
frame 4 Physics
INFO (I3Module): Opened file /cvmfs/icecube.opensciencegrid.org/data/i3-test-data/sim/corsika.F2K010001_IC59_slim.i3.gz (I3Reader.cxx:180 in void I3Reader::OpenNextFile())
frame 5 DAQ
frame 6 DAQ
frame 7 DAQ
frame 8 DAQ
frame 9 DAQ
frame 10 DAQ
frame 11 DAQ
frame 12 DAQ
frame 13 DAQ
frame 14 DAQ
frame 15 DAQ
frame 16 DAQ
frame 17 DAQ
frame 18 DAQ
frame 19 DAQ
frame 20 DAQ
frame 21 DAQ
frame 22 DAQ
frame 23 DAQ
frame 24 DAQ
frame 25 DAQ
frame 26 DAQ
frame 27 DAQ
frame 28 DAQ
frame 29 DAQ
frame 30 DAQ
frame 31 DAQ
frame 32 DAQ
frame 33 DAQ
frame 34 DAQ
frame 35 DAQ
frame 36 DAQ
frame 37 DAQ
frame 38 DAQ
frame 39 DAQ
frame 40 DAQ
frame 41 DAQ
frame 42 DAQ
frame 43 DAQ
frame 44 DAQ
frame 45 DAQ
frame 46 DAQ
frame 47 DAQ
frame 48 DAQ
frame 49 DAQ
frame 50 DAQ
frame 51 DAQ
frame 52 DAQ
frame 53 DAQ
frame 54 DAQ
frame 55 DAQ
frame 56 DAQ
frame 57 DAQ
frame 58 DAQ
frame 59 DAQ
frame 60 DAQ
frame 61 DAQ
frame 62 DAQ
frame 63 DAQ
frame 64 DAQ
frame 65 DAQ
frame 66 DAQ
frame 67 DAQ
frame 68 DAQ
frame 69 DAQ
frame 70 DAQ
frame 71 DAQ
frame 72 DAQ
frame 73 DAQ
frame 74 DAQ
frame 75 DAQ
frame 76 DAQ
frame 77 DAQ
frame 78 DAQ
frame 79 DAQ
frame 80 DAQ
frame 81 DAQ
frame 82 DAQ
frame 83 DAQ
frame 84 DAQ
frame 85 DAQ
frame 86 DAQ
frame 87 DAQ
frame 88 DAQ
frame 89 DAQ
frame 90 DAQ
frame 91 DAQ
frame 92 DAQ
frame 93 DAQ
frame 94 DAQ
frame 95 DAQ
frame 96 DAQ
frame 97 DAQ
frame 98 DAQ
frame 99 DAQ
frame 100 DAQ
frame 101 DAQ
frame 102 DAQ
frame 103 DAQ
frame 104 DAQ
NOTICE (I3Tray): I3Tray finishing... (I3Tray.cxx:480 in void I3Tray::Execute(unsigned int))
.INFO (I3Module): Opened file /cvmfs/icecube.opensciencegrid.org/data/i3-test-data/event-viewer/Level3aGCD_IC79_EEData_Run00115990_slim.i3 (I3Reader.cxx:180 in void I3Reader::OpenNextFile())
Configure with filenamelist: /cvmfs/icecube.opensciencegrid.org/data/i3-test-data/event-viewer/Level3aGCD_IC79_EEData_Run00115990_slim.i3
frame 0 Geometry
frame 1 DAQ
frame 2 Physics
frame 3 DAQ
frame 4 Physics
frame 5 DAQ
frame 6 Physics
frame 7 DAQ
NOTICE (I3Tray): I3Tray finishing... (I3Tray.cxx:480 in void I3Tray::Execute(unsigned int))
.INFO (I3Module): Opened file /cvmfs/icecube.opensciencegrid.org/data/i3-test-data/event-viewer/Level3aGCD_IC79_EEData_Run00115990_slim.i3 (I3Reader.cxx:180 in void I3Reader::OpenNextFile())
Configure with filenamelist: /cvmfs/icecube.opensciencegrid.org/data/i3-test-data/event-viewer/Level3aGCD_IC79_EEData_Run00115990_slim.i3 /cvmfs/icecube.opensciencegrid.org/data/i3-test-data/event-viewer/Level3aGCD_IC79_EEData_Run00115990_slim.i3
frame 0 Geometry
frame 1 DAQ
frame 2 Physics
frame 3 DAQ
frame 4 Physics
frame 5 DAQ
frame 6 Physics
frame 7 DAQ
INFO (I3Module): Opened file /cvmfs/icecube.opensciencegrid.org/data/i3-test-data/event-viewer/Level3aGCD_IC79_EEData_Run00115990_slim.i3 (I3Reader.cxx:180 in void I3Reader::OpenNextFile())
frame 8 Geometry
shovelio keys: 2 I3Reader keys: 13
2 common keys
11 keys only in I3Reader: InIceRawData IceTopVEMPulsesSLC FilterMask IceTopRawData PoleMuonLlhFitFitParams PoleMuonLlhFit PoleMuonLinefit DrivingTime PoleMuonLinefitParams OfflinePulses I3TriggerHierarchy
ERROR (I3Module): All: Exception thrown (I3Module.cxx:116 in void I3Module::Do(void (I3Module::*)()))
F
======================================================================
FAIL: test_repeated_file (__main__.ShovelioI3ReaderComparisonTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "steamshovel/resources/resources/test/test_shovelio_against_I3Reader.py", line 134, in test_repeated_file
    self._run_on( [ f, f ] )
  File "steamshovel/resources/resources/test/test_shovelio_against_I3Reader.py", line 126, in _run_on
    self.fail(e.message)
AssertionError

----------------------------------------------------------------------
Ran 3 tests in 2.336s

FAILED (failures=1)
```

Migrated from https://code.icecube.wisc.edu/ticket/1719
```json
{
    "status": "closed", 
    "changetime": "2019-02-13T14:12:47", 
    "description": "There is apparently a race condition that needs fixing:\n\n{{{\ndschultz@tide2:~/Documents/offline_software/serialization/build_1.60$ python steamshovel/resources/resources/test/test_shovelio_against_I3Reader.py \nINFO (I3Module): Opened file /cvmfs/icecube.opensciencegrid.org/data/i3-test-data/sim/GCD.i3.gz (I3Reader.cxx:180 in void I3Reader::OpenNextFile())\nConfigure with filenamelist: /cvmfs/icecube.opensciencegrid.org/data/i3-test-data/sim/GCD.i3.gz /cvmfs/icecube.opensciencegrid.org/data/i3-test-data/sim/corsika.F2K010001_IC59_slim.i3.gz\nframe 1 Geometry\nframe 2 Calibration\nframe 3 DetectorStatus\nframe 4 Physics\nINFO (I3Module): Opened file /cvmfs/icecube.opensciencegrid.org/data/i3-test-data/sim/corsika.F2K010001_IC59_slim.i3.gz (I3Reader.cxx:180 in void I3Reader::OpenNextFile())\nframe 5 DAQ\nframe 6 DAQ\nframe 7 DAQ\nframe 8 DAQ\nframe 9 DAQ\nframe 10 DAQ\nframe 11 DAQ\nframe 12 DAQ\nframe 13 DAQ\nframe 14 DAQ\nframe 15 DAQ\nframe 16 DAQ\nframe 17 DAQ\nframe 18 DAQ\nframe 19 DAQ\nframe 20 DAQ\nframe 21 DAQ\nframe 22 DAQ\nframe 23 DAQ\nframe 24 DAQ\nframe 25 DAQ\nframe 26 DAQ\nframe 27 DAQ\nframe 28 DAQ\nframe 29 DAQ\nframe 30 DAQ\nframe 31 DAQ\nframe 32 DAQ\nframe 33 DAQ\nframe 34 DAQ\nframe 35 DAQ\nframe 36 DAQ\nframe 37 DAQ\nframe 38 DAQ\nframe 39 DAQ\nframe 40 DAQ\nframe 41 DAQ\nframe 42 DAQ\nframe 43 DAQ\nframe 44 DAQ\nframe 45 DAQ\nframe 46 DAQ\nframe 47 DAQ\nframe 48 DAQ\nframe 49 DAQ\nframe 50 DAQ\nframe 51 DAQ\nframe 52 DAQ\nframe 53 DAQ\nframe 54 DAQ\nframe 55 DAQ\nframe 56 DAQ\nframe 57 DAQ\nframe 58 DAQ\nframe 59 DAQ\nframe 60 DAQ\nframe 61 DAQ\nframe 62 DAQ\nframe 63 DAQ\nframe 64 DAQ\nframe 65 DAQ\nframe 66 DAQ\nframe 67 DAQ\nframe 68 DAQ\nframe 69 DAQ\nframe 70 DAQ\nframe 71 DAQ\nframe 72 DAQ\nframe 73 DAQ\nframe 74 DAQ\nframe 75 DAQ\nframe 76 DAQ\nframe 77 DAQ\nframe 78 DAQ\nframe 79 DAQ\nframe 80 DAQ\nframe 81 DAQ\nframe 82 DAQ\nframe 83 DAQ\nframe 84 DAQ\nframe 85 DAQ\nframe 86 DAQ\nframe 87 DAQ\nframe 88 DAQ\nframe 89 DAQ\nframe 90 DAQ\nframe 91 DAQ\nframe 92 DAQ\nframe 93 DAQ\nframe 94 DAQ\nframe 95 DAQ\nframe 96 DAQ\nframe 97 DAQ\nframe 98 DAQ\nframe 99 DAQ\nframe 100 DAQ\nframe 101 DAQ\nframe 102 DAQ\nframe 103 DAQ\nframe 104 DAQ\nNOTICE (I3Tray): I3Tray finishing... (I3Tray.cxx:480 in void I3Tray::Execute(unsigned int))\n.INFO (I3Module): Opened file /cvmfs/icecube.opensciencegrid.org/data/i3-test-data/event-viewer/Level3aGCD_IC79_EEData_Run00115990_slim.i3 (I3Reader.cxx:180 in void I3Reader::OpenNextFile())\nConfigure with filenamelist: /cvmfs/icecube.opensciencegrid.org/data/i3-test-data/event-viewer/Level3aGCD_IC79_EEData_Run00115990_slim.i3\nframe 0 Geometry\nframe 1 DAQ\nframe 2 Physics\nframe 3 DAQ\nframe 4 Physics\nframe 5 DAQ\nframe 6 Physics\nframe 7 DAQ\nNOTICE (I3Tray): I3Tray finishing... (I3Tray.cxx:480 in void I3Tray::Execute(unsigned int))\n.INFO (I3Module): Opened file /cvmfs/icecube.opensciencegrid.org/data/i3-test-data/event-viewer/Level3aGCD_IC79_EEData_Run00115990_slim.i3 (I3Reader.cxx:180 in void I3Reader::OpenNextFile())\nConfigure with filenamelist: /cvmfs/icecube.opensciencegrid.org/data/i3-test-data/event-viewer/Level3aGCD_IC79_EEData_Run00115990_slim.i3 /cvmfs/icecube.opensciencegrid.org/data/i3-test-data/event-viewer/Level3aGCD_IC79_EEData_Run00115990_slim.i3\nframe 0 Geometry\nframe 1 DAQ\nframe 2 Physics\nframe 3 DAQ\nframe 4 Physics\nframe 5 DAQ\nframe 6 Physics\nframe 7 DAQ\nINFO (I3Module): Opened file /cvmfs/icecube.opensciencegrid.org/data/i3-test-data/event-viewer/Level3aGCD_IC79_EEData_Run00115990_slim.i3 (I3Reader.cxx:180 in void I3Reader::OpenNextFile())\nframe 8 Geometry\nshovelio keys: 2 I3Reader keys: 13\n2 common keys\n11 keys only in I3Reader: InIceRawData IceTopVEMPulsesSLC FilterMask IceTopRawData PoleMuonLlhFitFitParams PoleMuonLlhFit PoleMuonLinefit DrivingTime PoleMuonLinefitParams OfflinePulses I3TriggerHierarchy\nERROR (I3Module): All: Exception thrown (I3Module.cxx:116 in void I3Module::Do(void (I3Module::*)()))\nF\n======================================================================\nFAIL: test_repeated_file (__main__.ShovelioI3ReaderComparisonTest)\n----------------------------------------------------------------------\nTraceback (most recent call last):\n  File \"steamshovel/resources/resources/test/test_shovelio_against_I3Reader.py\", line 134, in test_repeated_file\n    self._run_on( [ f, f ] )\n  File \"steamshovel/resources/resources/test/test_shovelio_against_I3Reader.py\", line 126, in _run_on\n    self.fail(e.message)\nAssertionError\n\n----------------------------------------------------------------------\nRan 3 tests in 2.336s\n\nFAILED (failures=1)\n}}}", 
    "reporter": "david.schultz", 
    "cc": "", 
    "resolution": "fixed", 
    "_ts": "1550067167842669", 
    "component": "combo core", 
    "summary": "[steamshovel] test failure - race condition", 
    "priority": "major", 
    "keywords": "", 
    "time": "2016-05-31T22:14:58", 
    "milestone": "", 
    "owner": "sander.vanheule", 
    "type": "defect"
}
```

**Comment by icecube-trac** [comment]
Trac update at 20160809T03:07:59: 
* olivas changed owner from "hdembinski" to "sander"
* olivas changed status from "new" to "assigned"

**Comment by icecube-trac** [comment]
Trac update at 20160815T17:04:49: olivas changed owner from "sander" to "sander.vanheule"

**Comment by icecube-trac** [comment]
Trac update at 20160818T16:37:26: 
* sander.vanheule changed status from "assigned" to "closed"
* sander.vanheule commented:

> In changeset 149355:
> 
> ```text
> #!CommitTicketReference repository="IceCube" revision="149355"
> Fix #1711: shovelio caches 'fast-read frame' with only the I3EventHeader present.
> ```


* sander.vanheule changed resolution from "" to "fixed"

**Comment by icecube-trac** [comment]
Trac update at 20160819T12:05:04: sander.vanheule commented:

> In changeset 149410:
> 
> ```text
> #!CommitTicketReference repository="IceCube" revision="149410"
> ```
> 
> When caching a frame with an uncached (fast-read) parent in the indexing thread, a parent merge is performed by the I3FrameMixer. Since the I3FrameMixer merges cached frame keys of the incomplete parent frame, the cached child frame is also incomplete. The code assumes however that a cached frame requires no merging, resulting in an incomplete frame being returned.
> 
> Commit changeset 149355 only worked around this problem by forcing a frame with an uncached parent to stay out of the cache.

**Comment by icecube-trac** [comment]
Trac update at 20181010T09:59:08: jvansanten commented:

> In changeset 166119:
> 
> ```text
> #!CommitTicketReference repository="IceCube" revision="166119"
> Recursively cache parents of parents when merging frames
> 
> This fixes a severe performance regression introduced in r149410, which
> attempted to fix #1711. As a side-effect, though, it only cached direct
> parents. This meant that the Q parent of a P frame would be cached, but GCD
> frames would be read from the input file for every P frame. For a gzipped
> Upgrade mDOM GCD, this caused steamshovel to take 30 seconds to move between
> frames.
> ```

**Comment by icecube-trac** [comment]
Trac update at 20190112T03:42:09: nwhitehorn commented:

> In changeset 170521:
> 
> ```text
> #!CommitTicketReference repository="IceCube" revision="170521"
> Recursively cache parents of parents when merging frames
> 
> This fixes a severe performance regression introduced in r149410, which
> attempted to fix #1711. As a side-effect, though, it only cached direct
> parents. This meant that the Q parent of a P frame would be cached, but GCD
> frames would be read from the input file for every P frame. For a gzipped
> Upgrade mDOM GCD, this caused steamshovel to take 30 seconds to move between
> frames.
> ```

**Comment by icecube-trac** [comment]
Trac update at 20190213T14:12:47: 
* olivas changed milestone from "June 2016" to ""
* olivas commented:

> Milestone June 2016 deleted
